sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -194,10 +195,22 @@ case class FileSourceScanExec(
     disableBucketedScan: Boolean = false)
   extends DataSourceScanExec {
 
+  lazy val outputMetadataStruct: Option[AttributeReference] =
+    output.collectFirst { case MetadataAttribute(attr) => attr }
```

Comment:
This is probably a design problem: do we prefer 4 flat columns or one struct-type column?

Reply:
"4 flat columns" was the original design. As per suggestions: changed from 4 flat columns to 4 fields under one struct:
- wrap in `_metadata`, potentially reduce name conflict with user's columns
- can query all available metadata information by selecting `_metadata`
- easy to maintain and extend (?)

I can see there's a clear disadvantage in doing struct-type: if users only select `_metadata.file_size`, we still need to fill all fields (hurt the performance). But it could be improved/fixed in follow-up PRs.